README.md
[![License: MPL 2.0][mpl-svg]][mpl] [![Test Status][travis-badge]][travis] [![Build Status][circleci-badge]][circleci] [![Connect to Matrix via the Riot webapp][matrix-badge]][matrix]

# Megaphone

## What is it?

Megaphone is an internal Mozilla system providing global broadcasts for Firefox.

Traditionally Firefox has polled multiple services at different frequencies (e.g. every 24 hours) to check for updates. Megaphone serves as an alternative, notifying user agents of new updates in near real time (within 5 minutes) over the [WebPush WebSocket protocol].

This enables Firefox to:

* [Revoke HTTPS Certificates] or [malicious extensions] immediately after security incidents occur
* Update quicker (Firefox itself, [tracking protection lists], or [general settings])
* Provide faster turn-around/feedback loops for studies/experiments ([Shield])

All via one unified, simpler client-side service that doesn't require polling.

This repository provides a Rust based Megaphone endpoint (API). Broadcasts are sent to the Megaphone endpoint. The [autopush-rs service] polls the endpoint as the source of truth for current broadcasts, ultimately delivering them to clients.

Also see the [API doc].


## Requirements

 * Rust nightly as specified in the `rust-toolchain` file (recognized by cargo) in the root of the project (recognized by cargo).
 * MySQL 5.7 (or compatible)
 * libmysqlclient (brew install mysql on macOS, apt-get install libmysqlclient-dev on Ubuntu)

 * *For running the docker image locally: docker-compose v1.21.0 or later

## Setting Up

1) [Install Rust]

2) Create a `megaphone` user/database

3) Run:

  $ export ROCKET_DATABASE_URL=mysql://scott:tiger@mydatabase/megaphone
  $ cargo run

## Running the Docker Image

1) [Install docker-compose]

2) From a dedicated screen (or tmux window)

$ `docker-compose up`

This will create two intertwined docker images:

***db_1*** - the database image. This image is a local test image. The database can be accessed via `mysql -umegaphone -ptest -h localhost --port 4306 megaphone`.

***app_1*** - the **megaphone** application. This is accessible via port 8000.


# API

Megaphone is normally called via a HTTP interface using Authorized calls. Responses are generally JSON objects with appropriate HTTP status codes to indicate success/failure.

## Authorization

All calls to Megaphone (minus the Dockerflow Status Checks) require authorization. Authorization is specified by the `Authorization` header via Bearer tokens specified in the application's configuration.

e.g.

```
export ROCKET_BROADCASTER_AUTH={test=["foobar"]}
export ROCKET_READER_AUTH={autopush=["quux"]}
```

The *test* broadcaster would specify:

```
Authorization: Bearer foobar
```

The *autopush* reader would specify:

```
Authorization: Bearer quux
```


## PUT /v1/broadcasts/< broadcaster_id > /< bchannel_id >

Broadcast a new version.

The body of the PUT request becomes the new version value.

A special version value of "____NOP____" (the string "NOP" prefixed and suffixed by four underscores) signals a "No Operation" to clients: that no action should take place, effectively overwriting and cancelling any pending version update.

The return value is a JSON structure including the HTTP status of the result: successful results either being a `201` code for newly created broadcasts or `200` for an update to an existing broadcast.

```javascript
{
   "code": 200
}
```


## GET /v1/broadcasts

Read the current broadcasts.

The return value is a JSON structure including the HTTP status of the result and a `broadcasts` object consisting of broadcastIDs and their current versions.

```javascript
{
   "code": 200,
   "broadcasts": {
      "test/broadcast1": "v3",
      "test/broadcast2": "v0"
   }
}
```

## Dockerflow Status Checks:

## GET /\_\_heartbeat__

Return the status of the server.

This call is only used for server status checks.


## GET /\_\_lbheartbeat__

Return a light weight status check (200 OK).

This call is only used for the Load Balancer's check.


## GET /\_\_version__

Return a JSON response of the version information of the server.


[mpl-svg]: https://img.shields.io/badge/License-MPL%202.0-blue.svg
[mpl]: https://opensource.org/licenses/MPL-2.0
[travis-badge]: https://travis-ci.org/mozilla-services/megaphone.svg?branch=master
[travis]: https://travis-ci.org/mozilla-services/megaphone
[circleci-badge]: https://circleci.com/gh/mozilla-services/megaphone.svg?style=shield&circle-token=074ae89011d1a7601378c41a4351e1e03f1e8177
[circleci]: https://circleci.com/gh/mozilla-services/megaphone
[matrix-badge]: https://img.shields.io/badge/chat%20on%20[m]-%23services%3Amozilla.org-blue
[matrix]: https://chat.mozilla.org/#/room/#services:mozilla.org

[WebPush WebSocket protocol]: https://mozilla-push-service.readthedocs.io/en/latest/design/#simplepush-protocol
[revoke HTTPS Certificates]: https://blog.mozilla.org/security/2015/03/03/revoking-intermediate-certificates-introducing-onecrl/
[malicious extensions]: https://wiki.mozilla.org/Blocklisting
[tracking protection lists]: https://wiki.mozilla.org/Security/Safe_Browsing
[general settings]: https://wiki.mozilla.org/Firefox/RemoteSettings
[Shield]: https://wiki.mozilla.org/Firefox/Shield/Shield_Studies
[autopush-rs service]: https://github.com/mozilla-services/autopush-rs
[API doc]: https://docs.google.com/document/d/1Wxqf1a4HDkKgHDIswPmhmdvk8KPoMEh2q6SPhaz4LNE

[Install Rust]: https://rustup.rs/
[Install docker-compose]: https://docs.docker.com/compose/install/


CHANGELOG.md
<a name=""></a>
##  (2023-08-29)


#### Chore

*   Update 2023-06 (#108) ([94dfd04b](https://github.com/mozilla-services/megaphone/commit/94dfd04b3793c496c4a0cacdee09b92543d1da42))
*   tag 0.2.0 (#98) ([8471abc6](https://github.com/mozilla-services/megaphone/commit/8471abc6c31ddc62f13ad8b05f30e477e22c0338))



<a name="0.2.0"></a>
## 0.2.0 (2021-09-28)


#### Bug Fixes

*   oops, improve the health check (#97) ([d26e6d23](https://github.com/mozilla-services/megaphone/commit/d26e6d23d7543d0c2925318907b6eed839b513f2), closes [#96](https://github.com/mozilla-services/megaphone/issues/96))
*   Add libcurl4 to Dockerfile ([ab19b4f7](https://github.com/mozilla-services/megaphone/commit/ab19b4f7e382e6002150218f9fffeaf556a79267), closes [#94](https://github.com/mozilla-services/megaphone/issues/94))

#### Features

*   Add metrics ([9fdb5973](https://github.com/mozilla-services/megaphone/commit/9fdb5973a05d49bfd3a49b8fc2c6ebc4f4c00182), closes [#83](https://github.com/mozilla-services/megaphone/issues/83))
*   add sentry error reporting ([ec2a104c](https://github.com/mozilla-services/megaphone/commit/ec2a104ccdc65749a457ab4e153f19842e18f593), closes [#70](https://github.com/mozilla-services/megaphone/issues/70))

#### Chore

*   Base to rocket "0.4" ([67f4b98f](https://github.com/mozilla-services/megaphone/commit/67f4b98fd4f4cc236d0ca3c5405c709e2ff182eb), closes [#84](https://github.com/mozilla-services/megaphone/issues/84))
*   update circleci to use new docker auth (#85) ([0813565e](https://github.com/mozilla-services/megaphone/commit/0813565eae8d6df8db816ada519dde320416762c))
*   Update Docker to rust 1.45 ([58d0cc4a](https://github.com/mozilla-services/megaphone/commit/58d0cc4a12e9a3a11d5ddd1d0d142f354e29baf4))
*   add a badge for the matrix channel ([40cacd24](https://github.com/mozilla-services/megaphone/commit/40cacd24a6ec3737906eb1cc96fbc0c1c7a83ca2))



<a name="0.1.6"></a>
## 0.1.6 (2020-02-29)


#### Chore

*   upgrade debian stretch -> buster ([85cbe9e8](https://github.com/mozilla-services/megaphone/commit/85cbe9e8cf8b27361519ed7bb1cf00d283e90aa5))
*   fix cargo install ([34b9809c](https://github.com/mozilla-services/megaphone/commit/34b9809cadce9cbc289781615f524bf844bdd44d))
*   cargo fmt ([1fd2e8e3](https://github.com/mozilla-services/megaphone/commit/1fd2e8e30e8276374a5d559192080b615684d806))
*   cargo fix --edition-idioms ([e2836bd7](https://github.com/mozilla-services/megaphone/commit/e2836bd70fe91ac0e06ec2bfb4fbbeda07107ddd))
*   cargo fix --edition ([372dc903](https://github.com/mozilla-services/megaphone/commit/372dc903e736d2aa37523ba0791eb62ee9008273))
*   update deps ([cb2a11e5](https://github.com/mozilla-services/megaphone/commit/cb2a11e5a1daa0f5012142cf094eabc989c3991b), closes [#76](https://github.com/mozilla-services/megaphone/issues/76))
*   update deps per cargo audit ([e6585acc](https://github.com/mozilla-services/megaphone/commit/e6585acccc2c8044e22c1246b98e95fa68fc6d97), closes [#74](https://github.com/mozilla-services/megaphone/issues/74))
*   utilize a rust-toolchain file ([a4b6e1f1](https://github.com/mozilla-services/megaphone/commit/a4b6e1f1155a13bc32dbcad1d06c04303666ea02))

#### Features

*   update dependencies ([1aa4aa1d](https://github.com/mozilla-services/megaphone/commit/1aa4aa1d233df749fbced6db720b0d19ee21df59), closes [#62](https://github.com/mozilla-services/megaphone/issues/62))
*   improve the README ([e1aa5077](https://github.com/mozilla-services/megaphone/commit/e1aa5077069f6b5a66eb04b7791c6510336cbc0f), closes [#59](https://github.com/mozilla-services/megaphone/issues/59))

#### Bug Fixes

*   get_bool now returns ConfigError::Missing ([644d33f4](https://github.com/mozilla-services/megaphone/commit/644d33f4c7d8575670ddff300925fe9f6dfb8387), closes [#67](https://github.com/mozilla-services/megaphone/issues/67))
*   propagate db pool errors into HandlerErrors ([3156643a](https://github.com/mozilla-services/megaphone/commit/3156643a338dbd99c42ce848a5c6e549a2f67324), closes [#59](https://github.com/mozilla-services/megaphone/issues/59))

#### Refactor

*   utilize slog_derive ([16b0922f](https://github.com/mozilla-services/megaphone/commit/16b0922f19cc4d9eb5e6c78243bc4cf0b653895a), closes [#64](https://github.com/mozilla-services/megaphone/issues/64))
*   prefer regular slog macro names ([1e6b61df](https://github.com/mozilla-services/megaphone/commit/1e6b61df8500e52b1296193e10ba7039d0b1dc62))



<a name="0.1.5"></a>
## 0.1.5 (2018-08-21)


#### Features

*   hacky support of ROCKET_LOG=off ([584f758b](https://github.com/mozilla-services/megaphone/commit/584f758bfe9ca3aec83d39848b3d441baa22b092), closes [#54](https://github.com/mozilla-services/megaphone/issues/54))



<a name="0.1.4"></a>
## 0.1.4 (2018-06-27)


#### Chore

*   include /app/version.json per dockerflow ([d5978e3c](https://github.com/mozilla-services/megaphone/commit/d5978e3c9a475208965537c21d88681277876890))

#### Features

*   warn log for ACL related errors ([109476a5](https://github.com/mozilla-services/megaphone/commit/109476a5e6313de761a593b5e4dcbd18a93f34c4), closes [#48](https://github.com/mozilla-services/megaphone/issues/48))
*   render a unique errno code per error ([57025884](https://github.com/mozilla-services/megaphone/commit/57025884993b16dfdf779d414ccda9afa353dc78), closes [#46](https://github.com/mozilla-services/megaphone/issues/46))
*   validate the PUT input ([6fbf572d](https://github.com/mozilla-services/megaphone/commit/6fbf572d685a32c7b60fbb0b7531422fbbd0781d), closes [#24](https://github.com/mozilla-services/megaphone/issues/24))
*   add a docker compose setup ([2c1aa8f1](https://github.com/mozilla-services/megaphone/commit/2c1aa8f1441d8b66eb23c8370c18806738fdc674), closes [#43](https://github.com/mozilla-services/megaphone/issues/43))

src/http.rs
// Include for clippy error on `fn lbhearbeat` expansion
#![allow(clippy::let_unit_value)]

use std::env;
use std::io::Read;
use std::time::Instant;

use diesel::RunQueryDsl;
use lazy_static::lazy_static;
use regex::Regex;
use rocket::{
    self,
    config::RocketConfig,
    data::{self, FromDataSimple},
    http::Status,
    outcome::IntoOutcome,
    request::{self, FromRequest},
    response::{content, status},
    Data,
    Outcome::{Failure, Success},
    Request, Rocket,
};
use rocket_contrib::{json, json::JsonValue};
use slog::{error, info};

use crate::auth;
use crate::db::{
    self,
    models::{Broadcaster, Reader},
};
use crate::error::{HandlerError, HandlerErrorKind, HandlerResult, VALIDATION_FAILED};
use crate::logging::{self, RequestLogger};
use crate::metrics::Metrics;
use crate::tags::Tags;

lazy_static! {
    static ref URLSAFE_B64_RE: Regex = Regex::new(r"^[A-Za-z0-9\-_]+$").unwrap();
}

impl<'a, 'r> FromRequest<'a, 'r> for Broadcaster {
    type Error = HandlerError;

    fn from_request(request: &'a Request<'r>) -> request::Outcome<Self, HandlerError> {
        auth::authorized_broadcaster(request).into_outcome(VALIDATION_FAILED)
    }
}

/// A new broadcast
#[derive(Debug)]
struct VersionInput {
    value: String,
}

impl FromDataSimple for VersionInput {
    type Error = HandlerError;

    fn from_data(_: &Request<'_>, data: Data) -> data::Outcome<Self, HandlerError> {
        let mut value = String::new();
        if let Err(_e) = data.open().read_to_string(&mut value) {
            return Failure((
                VALIDATION_FAILED,
                HandlerErrorKind::MissingVersionDataError.into(),
            ));
        };
        if value.is_empty() || value.len() > 200 || !value.is_ascii() {
            return Failure((
                VALIDATION_FAILED,
                HandlerErrorKind::InvalidVersionDataError.into(),
            ));
        }
        Success(VersionInput { value })
    }
}

impl<'a, 'r> FromRequest<'a, 'r> for Reader {
    type Error = HandlerError;

    fn from_request(request: &'a Request<'r>) -> request::Outcome<Self, HandlerError> {
        auth::authorized_reader(request).into_outcome(VALIDATION_FAILED)
    }
}

// REST Functions

#[allow(clippy::too_many_arguments)]
/// Set a version for a broadcaster / bchannel
#[put("/v1/broadcasts/<broadcaster_id>/<bchannel_id>", data = "<version>")]
fn broadcast(
    conn: HandlerResult<db::Conn>,
    log: RequestLogger,
    broadcaster: HandlerResult<Broadcaster>,
    broadcaster_id: String,
    bchannel_id: String,
    version: HandlerResult<VersionInput>,
    metrics: Metrics,
    base_tags: Tags,
) -> HandlerResult<status::Custom<JsonValue>> {
    let conn = conn?;

    if broadcaster_id.len() > 64 || !URLSAFE_B64_RE.is_match(&broadcaster_id) {
        Err(HandlerErrorKind::InvalidBroadcasterId)?
    }
    if bchannel_id.len() > 128 || !URLSAFE_B64_RE.is_match(&bchannel_id) {
        Err(HandlerErrorKind::InvalidBchannelId)?
    }

    let mut tags = base_tags;
    let version = version?.value;

    tags.tags
        .insert("broadcaster".to_owned(), broadcaster_id.clone());
    tags.tags
        .insert("channel_id".to_owned(), bchannel_id.clone());
    tags.tags.insert("version".to_owned(), version.clone());
    metrics.incr_with_tags("broadcast.cmd.update", Some(tags.clone()));

    let start = Instant::now();
    let created = broadcaster?.broadcast_new_version(&conn, &bchannel_id, &version)?;
    metrics.timer_with_tags(
        "broadcast.update",
        (Instant::now() - start).as_millis() as u64,
        Some(tags),
    );
    let status = if created { Status::Created } else { Status::Ok };
    info!(
        log,
        "Broadcast: {}/{} new version: {}",
        broadcaster_id,
        bchannel_id,
        &version;
        "code" => status.code
    );
    Ok(status::Custom(
        status,
        json!({
            "code": status.code
        }),
    ))
}

/// Dump the current version table
#[get("/v1/broadcasts")]
fn get_broadcasts(
    conn: HandlerResult<db::Conn>,
    reader: HandlerResult<Reader>,
    metrics: Metrics,
) -> HandlerResult<JsonValue> {
    metrics.incr("broadcast.cmd.dump");
    let conn = conn?;
    let start = Instant::now();
    let broadcasts = reader?.read_broadcasts(&conn)?;
    metrics.timer_with_tags(
        "broadcast.dump",
        (Instant::now() - start).as_millis() as u64,
        None,
    );

    Ok(json!({
        "code": 200,
        "broadcasts": broadcasts
    }))
}

#[get("/v1/err")]
fn log_check(
    _conn: HandlerResult<db::Conn>,
    log: RequestLogger,
) -> HandlerResult<content::Json<&'static str>> {
    info!(log, "Oh my!");
    error!(log, "Oh dear!");
    // Attempt to force a message through.
    // sentry::capture_message("Oh bother!", sentry::protocol::Level::Info);
    // panic!("Oh noes!")
    Err(HandlerErrorKind::TestError.into())
}

#[get("/__version__")]
fn version() -> content::Json<&'static str> {
    content::Json(include_str!("../version.json"))
}

#[get("/__heartbeat__")]
fn heartbeat(conn: HandlerResult<db::Conn>, log: RequestLogger) -> status::Custom<JsonValue> {
    let result = conn.and_then(|conn| Ok(diesel::sql_query("SELECT 1").execute(&*conn)?));

    let status = match result {
        Ok(_) => Status::Ok,
        Err(e) => {
            let status = Status::ServiceUnavailable;
            error!(log, "Database heartbeat failed: {}", e; "code" => status.code);
            status
        }
    };

    let msg = if status == Status::Ok { "ok" } else { "error" };
    status::Custom(
        status,
        json!({
            "status": msg,
            "code": status.code,
            "database": msg,
        }),
    )
}

#[get("/__lbheartbeat__")]
fn lbheartbeat() {}

#[catch(404)]
fn not_found() -> HandlerResult<()> {
    Err(HandlerErrorKind::NotFound)?
}

pub fn get_sentry(config: &rocket::config::Config) -> Option<sentry::ClientInitGuard> {
    let opts = sentry::ClientOptions {
        // debug: true,
        ..Default::default()
    };
    if let Ok(sentry_dsn) = config.get_string("sentry_dsn") {
        Some(sentry::init((sentry_dsn, opts)))
    } else {
        // Check the global env to see if we need to connect to sentry.
        if env::var("SENTRY_DSN").is_ok() {
            Some(sentry::init(opts))
        } else {
            None
        }
    }
}

pub fn rocket() -> HandlerResult<Rocket> {
    // RocketConfig::init basically
    let rconfig = RocketConfig::read().unwrap_or_else(|_| {
        let path = env::current_dir()
            .unwrap()
            .join(format!(".{}.{}", "default", "Rocket.toml"));
        RocketConfig::active_default_from(Some(&path)).unwrap()
    });

    // rocket::ignite basically
    let config = rconfig.active().clone();
    setup_rocket(rocket::custom(config))
}

fn setup_rocket(rocket: Rocket) -> HandlerResult<Rocket> {
    let pool = db::pool_from_config(rocket.config())?;
    let authenticator = auth::BearerTokenAuthenticator::from_config(rocket.config())?;
    let environment = rocket.config().environment;
    let sentry_client = get_sentry(rocket.config());
    let logger = logging::init_logging(rocket.config(), &sentry_client)?;
    let tags = Tags::init(rocket.config())?;
    let metrics = Metrics::init(rocket.config(), &sentry_client)?;
    info!(logger, "Starting up");
    db::run_embedded_migrations(rocket.config())?;
    Ok(rocket
        .manage(pool)
        .manage(authenticator)
        .manage(environment)
        .manage(logger)
        .manage(metrics)
        .manage(tags)
        .manage(sentry_client)
        .mount(
            "/",
            routes![
                broadcast,
                get_broadcasts,
                version,
                heartbeat,
                lbheartbeat,
                log_check
            ],
        )
        .register(catchers![not_found]))
}

#[cfg(test)]
mod test {
    use crate::auth::test::to_table;
    use rocket::config::{Config, Environment, RocketConfig, Value as RValue};
    use rocket::http::{Header, Status};
    use rocket::local::Client;
    use rocket::response::Response;
    use rocket_contrib::json;
    use serde_json::{self, Value};

    use super::setup_rocket;

    /// Test auth headers
    enum Auth {
        Foo,
        FooAlt,
        Baz,
        Reader,
    }

    impl From<Auth> for Header<'static> {
        fn from(auth: Auth) -> Header<'static> {
            let token = match auth {
                Auth::Foo => "feedfacedeadbeef",
                Auth::FooAlt => "deadbeeffacefeed",
                Auth::Baz => "baada555deadbeef",
                Auth::Reader => "00000000deadbeef",
            };
            Header::new("Authorization".to_string(), format!("Bearer {}", token))
        }
    }

    /// Return a Rocket Client for testing
    ///
    /// The managed db pool is set to a maxiumum of one connection w/
    /// a transaction began that is never committed
    fn rocket_client() -> Client {
        // create a separate test config but inheriting database_url
        let rconfig = RocketConfig::read().expect("reading rocket Config failed");
        let database_url = rconfig
            .active()
            .get_str("database_url")
            .expect("ROCKET_DATABASE_URL undefined")
            .to_owned();
        let config = Config::build(Environment::Development)
            .extra("database_url", RValue::String(database_url))
            .extra("database_pool_max_size", 1)
            .extra("database_use_test_transactions", true)
            .extra("json_logging", false)
            .extra(
                "broadcaster_auth",
                to_table(
                    [
                        "foo=feedfacedeadbeef,deadbeeffacefeed",
                        "baz=baada555deadbeef",
                    ]
                    .to_vec(),
                ),
            )
            .extra(
                "reader_auth",
                to_table(["reader=00000000deadbeef"].to_vec()),
            )
            .unwrap();
        dbg!(&config);

        let rocket = setup_rocket(rocket::custom(config)).expect("rocket failed");
        Client::new(rocket).expect("rocket launch failed")
    }

    fn json_body(response: &mut Response<'_>) -> Value {
        assert!(response.content_type().map_or(false, |ct| ct.is_json()));
        serde_json::from_str(&response.body_string().unwrap()).unwrap()
    }

    #[test]
    fn test_put() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::Foo)
            .body("v0")
            .dispatch();
        assert_eq!(response.status(), Status::Created);
        assert_eq!(json_body(&mut response), *json!({"code": 201}));
        let mut response = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::FooAlt)
            .body("v1")
            .dispatch();
        assert_eq!(response.status(), Status::Ok);
        assert_eq!(json_body(&mut response), *json!({"code": 200}));
    }

    #[test]
    fn test_put_no_body() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::Foo)
            .dispatch();
        assert_eq!(response.status(), Status::BadRequest);
        let result = json_body(&mut response);
        assert_eq!(result["code"], Status::BadRequest.code);
        assert!(result["error"].as_str().unwrap().contains("Version"));
    }

    #[test]
    fn test_put_no_id() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo")
            .header(Auth::Foo)
            .body("v1")
            .dispatch();
        assert_eq!(response.status(), Status::NotFound);
        assert_eq!(
            json_body(&mut response),
            *json!({"code": 404, "errno": 123, "error": "Not Found"})
        );
    }

    #[test]
    fn test_put_no_auth() {
        let client = rocket_client();
        let mut response = client.put("/v1/broadcasts/foo/bar").body("v1").dispatch();
        assert_eq!(response.status(), Status::Unauthorized);
        assert!(response
            .headers()
            .get_one("WWW-Authenticate")
            .unwrap()
            .starts_with("Bearer "));
        let result = json_body(&mut response);
        assert_eq!(result["code"], 401);
    }

    #[test]
    fn test_put_bad_auth() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::Baz)
            .body("v1")
            .dispatch();
        assert_eq!(response.status(), Status::Forbidden);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 403);
    }

    #[test]
    fn test_put_bad_ids() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo+bar/baz")
            .header(Auth::Baz)
            .body("v1")
            .dispatch();
        assert_eq!(response.status(), Status::BadRequest);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 400);
        assert!(result["error"].as_str().unwrap().contains("broadcasterID"));

        let mut response = client
            .put("/v1/broadcasts/foo/bar+baz")
            .header(Auth::Baz)
            .body("v1")
            .dispatch();
        assert_eq!(response.status(), Status::BadRequest);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 400);
        assert!(result["error"].as_str().unwrap().contains("bchannelID"));
    }

    #[test]
    fn test_put_bad_version() {
        let client = rocket_client();
        let mut response = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::Baz)
            .body("v1".repeat(101))
            .dispatch();
        assert_eq!(response.status(), Status::BadRequest);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 400);
        assert!(result["error"].as_str().unwrap().contains("Version"));
    }

    #[test]
    fn test_get_no_auth() {
        let client = rocket_client();
        let mut response = client.get("/v1/broadcasts").dispatch();
        assert_eq!(response.status(), Status::Unauthorized);
        assert!(response
            .headers()
            .get_one("WWW-Authenticate")
            .unwrap()
            .starts_with("Bearer "));
        let result = json_body(&mut response);
        assert_eq!(result["code"], 401);
    }

    #[test]
    fn test_get_bad_auth() {
        let client = rocket_client();
        let mut response = client.get("/v1/broadcasts").header(Auth::Foo).dispatch();
        assert_eq!(response.status(), Status::Forbidden);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 403);
    }

    #[test]
    fn test_put_get() {
        let client = rocket_client();
        let _ = client
            .put("/v1/broadcasts/foo/bar")
            .header(Auth::FooAlt)
            .body("v1")
            .dispatch();
        let _ = client
            .put("/v1/broadcasts/baz/quux")
            .header(Auth::Baz)
            .body("v0")
            .dispatch();
        let mut response = client.get("/v1/broadcasts").header(Auth::Reader).dispatch();
        assert_eq!(response.status(), Status::Ok);
        assert_eq!(
            json_body(&mut response),
            *json!({"code": 200, "broadcasts": {"baz/quux": "v0", "foo/bar": "v1"}})
        );
    }

    #[test]
    fn test_version() {
        let client = rocket_client();
        let mut response = client.get("/__version__").dispatch();
        assert_eq!(response.status(), Status::Ok);
        let result = json_body(&mut response);
        assert_eq!(result["version"], "devel");
    }

    #[test]
    fn test_heartbeat() {
        let client = rocket_client();
        let mut response = client.get("/__heartbeat__").dispatch();
        assert_eq!(response.status(), Status::Ok);
        let result = json_body(&mut response);
        assert_eq!(result["code"], 200);
        assert_eq!(result["database"], "ok");
    }

    #[test]
    fn test_lbheartbeat() {
        let client = rocket_client();
        let mut response = client.get("/__lbheartbeat__").dispatch();
        assert_eq!(response.status(), Status::Ok);
        assert!(response.body().is_none());
    }
}
